### DISCUSSION TITLE
How to use update() with toggle button connected to checkbox changes?

### DISCUSSION THREAD

**Discussion by Anindya088** [zauberzeug/nicegui#1209] [Q&A] [answered]
### Question

Hi All,

I am trying to create 2 lists basically. Let's say first I import an excel file, and I have a list (A) of all sheet names.
Now I use some logic to decide some of them will not be shown, so make a smaller list (B)

So now what I am doing is I am using list B to show a toggle button. Idea is user press the button (which is sheet name), then the sheet data will be shown and next steps, etc.

Then below I am also showing the list A as a checkbox list to show which all sheets are currently active (meaning being displayed in the toggle button) versus which ones are turned off.

So, I want to build a connection between these 2, such that if you enable the checkbox below, that sheet name is auto-updated in the toggle button list. So now that sheet is "activated", so to say.

I am sharing my code below. I have managed to make it work, although probably not the best/efficient way, seems pretty clumsy to me. One of the thing I could not figure out is... is there a way to use the ```.update()``` property to update the toggle button source list?

Meaning I declare the toggle button first with the list B. For the checkbox set, I give a function in ```on_change=``` which will update list B and call for ```update()``` for the toggle button. 

But I could not make it work, instead, what I did was to declare a container up front, then in the ```on_change=``` function I cleared the container and created the toggle button again with the updated list B. This is probably not the best way to do this, so please let me know if there is an easier/more efficient way of refreshing/updating the toggle button connected to the checkbox changes. Actually I am doing few work items like this, so if I can get the thinking right on this one, hopefully I can solve the others also. 

Thanks a lot for all the guidance on this. I am sharing my code below here:

```python
# Excel file for media data
data = pd.read_excel('test_excel.xlsm', sheet_name=None, header=None)

# Keywords which if present, we are assuming these are not data sheets
sheet_exclude = ['sample', 'example', 'specimen'] 

# Create 2 lists, one for import flag (checkbox) another for display (toggle)
sheet_import = [(sheet, False if any(exclude_str in sheet.lower() for exclude_str in sheet_exclude) else True) for sheet in data.keys()]
sheet_display = [sheet for sheet, condition in sheet_import if condition is True]

# Initial Toggle Button
with ui.row() as buttons:
    toggle = ui.toggle(sheet_display)

# List for sheet selection
i=0
with ui.column().classes('border space-y-[1px] p-2'):
    for x,y in sheet_import:
      i=i+1
      ui.checkbox(x,value=y, on_change=lambda x=x,i=i: update(x,i)).props('dense size="xs"')

def update(value, count):
    sheet_import[count-1] = (value, not sheet_import[count-1][1])
    sheet_display = [sheet for sheet, condition in sheet_import if condition is True]

    buttons.clear()
    with buttons:
        toggle = ui.toggle(sheet_display)
```

**Comment by falkoschindler** [comment, marked as answer]
Hi @Anindya088,
I think your code can be boiled down to something like this:

```py
data = {'one': ..., 'two': ..., 'three': ..., 'sample': ..., 'example': ..., 'specimen': ...}

toggle = ui.toggle([name for name in data if not any(s in name.lower() for s in ['sample', 'example', 'specimen'])])

with ui.column().classes('border space-y-[1px] p-2'):
    for name in data:
        ui.checkbox(name, value=name in toggle.options, on_change=lambda name=name: check(name)).props('dense size=xs')

def check(name: str) -> None:
    if name in toggle.options:
        toggle.options.remove(name)
    else:
        toggle.options.append(name)
    toggle.update()
```

Since I don't have your Excel file, I improvised some dummy data. For discussions like this it is better anyway to avoid dependencies.

I found it simpler to just work with the toggle options directly instead of having keeping track of different lists. But the main takeaway is that you can change the options list, as long as you call `update()` afterwards.

**Reply by Anindya088** [reply]
Hi @falkoschindler , 

Thanks for the detailed help, and sorry for the late reply, I was offline for a few days.

I am trying out your advice and certainly looks more efficient that what I was trying out. I was not aware of the ```toggle.options.remove``` or ```toggle.options.append``` methods, I think that is why my ```update()``` was not working.

Thanks so much for explaining this, I will implement this for my case :)

Regards,
Anindya
